File: xeg-upgrader/dotnet/src/XegUpgrader/SummaryPrinter.cs
using System.Globalization;
using System.Text.Json;

namespace XegUpgrader;

public sealed record ParseSummary(
    string Filename,
    int Individuals,
    int Relationships,
    int Eggs,
    int Diseases,
    int Events,
    IReadOnlyList<string> DiseaseNames,
    IReadOnlyList<string> Warnings);

/// <summary>
/// Pure helpers that turn a parse-mode response into a human-readable
/// summary and inspect it for inconsistencies worth flagging.  No I/O.
/// </summary>
public static class SummaryPrinter
{
    private const int CountColumnWidth = 14;

    public static ParseSummary Summarise(JsonElement parseResponse, string filename)
    {
        var individuals = ArrayOf(parseResponse, "individuals");
        var relationships = ArrayOf(parseResponse, "relationships");
        var eggs = ArrayOf(parseResponse, "eggs");
        var diseases = ArrayOf(parseResponse, "diseases");

        return new ParseSummary(
            Filename: filename,
            Individuals: individuals.Count,
            Relationships: relationships.Count,
            Eggs: eggs.Count,
            Diseases: diseases.Count,
            Events: CountEvents(individuals) + CountEvents(relationships) + CountEvents(eggs),
            DiseaseNames: CollectDiseaseNames(diseases),
            Warnings: DetectWarnings(individuals, eggs, diseases));
    }

    public static string Render(ParseSummary summary, RunMode mode)
    {
        var writer = new StringWriter { NewLine = "\n" };
        writer.WriteLine($"File: {summary.Filename}");
        writer.WriteLine($"Mode: {ModeLine(mode)}");
        writer.WriteLine();
        writer.WriteLine("Counts");
        WriteCount(writer, "individuals", summary.Individuals);
        WriteCount(writer, "relationships", summary.Relationships);
        WriteCount(writer, "eggs", summary.Eggs);
        WriteCount(writer, "diseases", summary.Diseases);
        WriteCount(writer, "events", summary.Events);
        writer.WriteLine();
        writer.WriteLine("Diseases");
        WriteList(writer, summary.DiseaseNames);
        writer.WriteLine();
        writer.WriteLine("Warnings");
        WriteList(writer, summary.Warnings);
        return writer.ToString();
    }

    private static string ModeLine(RunMode mode) => mode switch
    {
        RunMode.Preview => "preview (no pedigree created)",
        RunMode.Create => "create (pedigree imported)",
        _ => throw new ArgumentOutOfRangeException(nameof(mode)),
    };

    private static void WriteCount(TextWriter writer, string label, int count)
    {
        var padded = label.PadRight(CountColumnWidth);
        writer.WriteLine($"  {padded} {count.ToString(CultureInfo.InvariantCulture)}");
    }

    private static void WriteList(TextWriter writer, IReadOnlyList<string> items)
    {
        if (items.Count == 0)
        {
            writer.WriteLine("  (none)");
            return;
        }
        foreach (var item in items)
        {
            writer.WriteLine($"  - {item}");
        }
    }

    private static IReadOnlyList<JsonElement> ArrayOf(JsonElement container, string key)
    {
        if (container.ValueKind != JsonValueKind.Object)
        {
            return Array.Empty<JsonElement>();
        }
        if (!container.TryGetProperty(key, out var value) || value.ValueKind != JsonValueKind.Array)
        {
            return Array.Empty<JsonElement>();
        }
        var list = new List<JsonElement>(value.GetArrayLength());
        foreach (var item in value.EnumerateArray())
        {
            list.Add(item);
        }
        return list;
    }

    private static int CountEvents(IReadOnlyList<JsonElement> containers)
    {
        var total = 0;
        foreach (var container in containers)
        {
            if (container.ValueKind != JsonValueKind.Object)
            {
                continue;
            }
            if (container.TryGetProperty("events", out var events) && events.ValueKind == JsonValueKind.Array)
            {
                total += events.GetArrayLength();
            }
        }
        return total;
    }

    private static List<string> CollectDiseaseNames(IReadOnlyList<JsonElement> diseases)
    {
        var names = new List<string>(diseases.Count);
        foreach (var disease in diseases)
        {
            names.Add(DisplayNameOf(disease) ?? "(unnamed)");
        }
        return names;
    }

    private static List<string> DetectWarnings(
        IReadOnlyList<JsonElement> individuals,
        IReadOnlyList<JsonElement> eggs,
        IReadOnlyList<JsonElement> diseases)
    {
        var warnings = new List<string>();

        var unknownSex = CountWhere(individuals, HasNoBiologicalSex);
        if (unknownSex > 0)
        {
            warnings.Add($"{unknownSex} individual(s) with unknown biological sex");
        }

        var unnamed = CountWhere(individuals, HasNoDisplayName);
        if (unnamed > 0)
        {
            warnings.Add($"{unnamed} individual(s) without a display name");
        }

        var orphanedEggs = CountWhere(eggs, HasNoRelationship);
        if (orphanedEggs > 0)
        {
            warnings.Add($"{orphanedEggs} egg(s) with no resolvable parent relationship");
        }

        var knownDiseaseIds = CollectDiseaseIds(diseases);
        var danglingManifestations = 0;
        foreach (var individual in individuals)
        {
            danglingManifestations += CountDanglingManifestations(individual, knownDiseaseIds);
        }
        if (danglingManifestations > 0)
        {
            warnings.Add($"{danglingManifestations} disease manifestation(s) with unknown disease_id");
        }

        return warnings;
    }

    private static int CountWhere(IReadOnlyList<JsonElement> items, Func<JsonElement, bool> predicate)
    {
        var count = 0;
        foreach (var item in items)
        {
            if (predicate(item))
            {
                count += 1;
            }
        }
        return count;
    }

    private static bool HasNoBiologicalSex(JsonElement individual)
    {
        if (individual.ValueKind != JsonValueKind.Object)
        {
            return false;
        }
        if (!individual.TryGetProperty("biological_sex", out var value))
        {
            return true;
        }
        if (value.ValueKind == JsonValueKind.Null)
        {
            return true;
        }
        if (value.ValueKind == JsonValueKind.String)
        {
            var text = value.GetString();
            return string.IsNullOrEmpty(text) || text.Equals("unknown", StringComparison.OrdinalIgnoreCase);
        }
        return false;
    }

    private static bool HasNoDisplayName(JsonElement individual)
    {
        return string.IsNullOrWhiteSpace(DisplayNameOf(individual));
    }

    private static bool HasNoRelationship(JsonElement egg)
    {
        if (egg.ValueKind != JsonValueKind.Object)
        {
            return true;
        }
        if (!egg.TryGetProperty("relationship_id", out var value))
        {
            return true;
        }
        return value.ValueKind == JsonValueKind.Null
            || (value.ValueKind == JsonValueKind.String && string.IsNullOrEmpty(value.GetString()));
    }

    private static HashSet<string> CollectDiseaseIds(IReadOnlyList<JsonElement> diseases)
    {
        var ids = new HashSet<string>(StringComparer.Ordinal);
        foreach (var disease in diseases)
        {
            if (disease.ValueKind != JsonValueKind.Object)
            {
                continue;
            }
            if (disease.TryGetProperty("id", out var id) && id.ValueKind == JsonValueKind.String)
            {
                var text = id.GetString();
                if (!string.IsNullOrEmpty(text))
                {
                    ids.Add(text);
                }
            }
        }
        return ids;
    }

    private static int CountDanglingManifestations(JsonElement individual, HashSet<string> knownIds)
    {
        if (individual.ValueKind != JsonValueKind.Object)
        {
            return 0;
        }
        if (!individual.TryGetProperty("diseases", out var diseases)
            || diseases.ValueKind != JsonValueKind.Array)
        {
            return 0;
        }
        var dangling = 0;
        foreach (var disease in diseases.EnumerateArray())
        {
            if (disease.ValueKind != JsonValueKind.Object)
            {
                continue;
            }
            if (!disease.TryGetProperty("disease_id", out var id) || id.ValueKind != JsonValueKind.String)
            {
                dangling += 1;
                continue;
            }
            var text = id.GetString();
            if (string.IsNullOrEmpty(text) || !knownIds.Contains(text))
            {
                dangling += 1;
            }
        }
        return dangling;
    }

    private static string? DisplayNameOf(JsonElement element)
    {
        if (element.ValueKind != JsonValueKind.Object)
        {
            return null;
        }
        if (!element.TryGetProperty("display_name", out var value) || value.ValueKind != JsonValueKind.String)
        {
            return null;
        }
        return value.GetString();
    }
}

File: xeg-upgrader/python/src/xeg_upgrader/summary_printer.py
"""Pure renderer for parse-mode responses into a human-readable summary.

Takes the JSON payload the Evagene API returns for
``POST /api/pedigrees/{id}/import/xeg?mode=parse`` and produces counts,
the list of diseases recovered, and any data-level warnings worth
flagging before a clinician commits to the import.
"""

from __future__ import annotations

from dataclasses import dataclass
from io import StringIO
from typing import Any

from .config import RunMode

_COUNT_COLUMN_WIDTH = 14


@dataclass(frozen=True)
class ParseSummary:
    filename: str
    individuals: int
    relationships: int
    eggs: int
    diseases: int
    events: int
    disease_names: tuple[str, ...]
    warnings: tuple[str, ...]


def summarise(parse_response: dict[str, Any], filename: str) -> ParseSummary:
    individuals = _array(parse_response, "individuals")
    relationships = _array(parse_response, "relationships")
    eggs = _array(parse_response, "eggs")
    diseases = _array(parse_response, "diseases")

    return ParseSummary(
        filename=filename,
        individuals=len(individuals),
        relationships=len(relationships),
        eggs=len(eggs),
        diseases=len(diseases),
        events=_count_events(individuals) + _count_events(relationships) + _count_events(eggs),
        disease_names=_collect_disease_names(diseases),
        warnings=_detect_warnings(individuals, eggs, diseases),
    )


def render(summary: ParseSummary, mode: RunMode) -> str:
    sink = StringIO()
    sink.write(f"File: {summary.filename}\n")
    sink.write(f"Mode: {_mode_line(mode)}\n\n")
    sink.write("Counts\n")
    _write_count(sink, "individuals", summary.individuals)
    _write_count(sink, "relationships", summary.relationships)
    _write_count(sink, "eggs", summary.eggs)
    _write_count(sink, "diseases", summary.diseases)
    _write_count(sink, "events", summary.events)
    sink.write("\nDiseases\n")
    _write_list(sink, summary.disease_names)
    sink.write("\nWarnings\n")
    _write_list(sink, summary.warnings)
    return sink.getvalue()


def _mode_line(mode: RunMode) -> str:
    if mode is RunMode.PREVIEW:
        return "preview (no pedigree created)"
    return "create (pedigree imported)"


def _write_count(sink: StringIO, label: str, count: int) -> None:
    sink.write(f"  {label.ljust(_COUNT_COLUMN_WIDTH)} {count}\n")


def _write_list(sink: StringIO, items: tuple[str, ...]) -> None:
    if not items:
        sink.write("  (none)\n")
        return
    for item in items:
        sink.write(f"  - {item}\n")


def _array(container: dict[str, Any], key: str) -> list[dict[str, Any]]:
    value = container.get(key)
    if not isinstance(value, list):
        return []
    return [item for item in value if isinstance(item, dict)]


def _count_events(containers: list[dict[str, Any]]) -> int:
    total = 0
    for container in containers:
        events = container.get("events")
        if isinstance(events, list):
            total += len(events)
    return total


def _collect_disease_names(diseases: list[dict[str, Any]]) -> tuple[str, ...]:
    return tuple(_display_name_of(d) or "(unnamed)" for d in diseases)


def _detect_warnings(
    individuals: list[dict[str, Any]],
    eggs: list[dict[str, Any]],
    diseases: list[dict[str, Any]],
) -> tuple[str, ...]:
    warnings: list[str] = []

    unknown_sex = sum(1 for ind in individuals if _has_no_biological_sex(ind))
    if unknown_sex:
        warnings.append(f"{unknown_sex} individual(s) with unknown biological sex")

    unnamed = sum(1 for ind in individuals if not _display_name_of(ind))
    if unnamed:
        warnings.append(f"{unnamed} individual(s) without a display name")

    orphaned_eggs = sum(1 for egg in eggs if not egg.get("relationship_id"))
    if orphaned_eggs:
        warnings.append(f"{orphaned_eggs} egg(s) with no resolvable parent relationship")

    known_ids = {d.get("id") for d in diseases if isinstance(d.get("id"), str)}
    dangling = sum(
        _count_dangling_manifestations(ind, known_ids) for ind in individuals
    )
    if dangling:
        warnings.append(f"{dangling} disease manifestation(s) with unknown disease_id")

    return tuple(warnings)


def _has_no_biological_sex(individual: dict[str, Any]) -> bool:
    value = individual.get("biological_sex")
    if value is None:
        return True
    return isinstance(value, str) and (not value or value.lower() == "unknown")


def _count_dangling_manifestations(
    individual: dict[str, Any],
    known_ids: set[Any],
) -> int:
    manifestations = individual.get("diseases")
    if not isinstance(manifestations, list):
        return 0
    dangling = 0
    for entry in manifestations:
        if not isinstance(entry, dict):
            continue
        disease_id = entry.get("disease_id")
        if not isinstance(disease_id, str) or disease_id not in known_ids:
            dangling += 1
    return dangling


def _display_name_of(element: dict[str, Any]) -> str:
    value = element.get("display_name")
    return value if isinstance(value, str) else ""
